### ISSUE TITLE
Sorting works inconsistently , when table column contains null/undefined/empty values

### ISSUE THREAD

**Issue by NiveditaPastor** [angular-ui/ng-grid#3360]
I observed that the handlenulls method present in all sort methods(exmaple : alphasort) is working inconsistently. It is unable to handle the empty values in present in the table while sorting.
Note - I am using Sort Ascending and Sort Descending options and not the header options.

Scenario -  When a column containing null values is sorted -

1) Step 1 : Ascending order - null values are pushed at the bottom, values gets sorted correctly
2) Step 2 : Descending order - nothing happens ,sorting does not work.

**Comment by PaulL1** [comment]
Are there any errors on the javascript console when it fails to work?  Are you able to provide a plunker?

**Comment by NiveditaPastor** [comment]
Here I have made the changes and added empty rows in the tables. Now when you will sort using "Sort Ascending" or "Sort Descending" drop down actions , they are not working as expected. Please let me know if more information is needed.

http://plnkr.co/edit/Vpdn95eOUA5AD4If53FY?p=preview

**Comment by NiveditaPastor** [comment]
@PaulL1  Hey Paul , did you get time to look at this issue. I have provided plunker example above.

**Comment by techieanant** [comment]
I've been watching this issue for months now. Do we have a fix for this yet ?

**Comment by PaulL1** [comment]
No, I haven't had time to look at it, but may be able to do so soon.

**Comment by aalesar** [comment]
I have the same issue.

**Comment by bkmagnetron** [comment]
I also have the same issue.

**Comment by patimanas321** [comment]
I too have the same issue.

**Comment by miseeger** [comment]
Using ui-grid v3.2.6 as of 2016-07-14, I created a quick workaround for this (only tested with r/o date columns) and created a custom Filter:

```javascript
(function() {
'use strict';

    angular
        .module('app')
        .filter('nullDate', nullDateFilter);
    
    nullDateFilter.$inject = ['dateFilter']
    function nullDateFilter(dateFilter) {
        return function(input, pattern) {
            if (input === null){
                return ' '; // space: #32
            } else {
                return dateFilter(input, pattern)
            }
        };
    };

})();
```
The nullDateFilter injects the dateFilter from angular and uses it so it's possible to pass a date pattern.

After that, you just use this filter in the column defs and switch on the sortCellFiltered function. This does the sorting regarding the set cellFilter. 

`{ name: 'myNullableDateField', displayName: 'Date', cellFilter: 'nullDate: "dd.MM.yyyy"', sortCellFiltered: true }`

The sorting now behaves as supposed ("nulls" on top when sorting ascending) and it works also if you have a custom celltemplate defined. With little changes, you can also apply a custom filter for nullable strings.

Please have a try, yourselves. I hope it helps :smiley:

**Comment by miseeger** [comment]
I ran into the same problem using ui-grid v3.2.6 as of 2016-07-14. So I created a quick workaround for this. It uses the `sortingAlgorithm` option (on a clumn def) which refers to a custom sort function.

For my nullable date columns:

```javascript
function sortNullDate(a, b) {
    a = a === null ? '1899-12-31 23:59:59' : a; 
    b = b === null ? '1899-12-31 23:59:59' : b;
    if (a === b) { return 0 }; 
    if (a < b) { return -1 }; 
    if (a > b) { return 1 }; 
};
```

For my empty-string columns:

```javascript
function sortNullString(a, b) {
    a = a == '' ? ' ' : a; 
    b = b == '' ? ' ' : b;
    if (a === b) { return 0 }; 
    if (a < b) { return -1 }; 
    if (a > b) { return 1 }; 
};
```

Usage:

`{ name: 'myNullableDateField', displayName: 'Date', cellFilter: 'nullDate: "dd.MM.yyyy"', sortAlgorithm: sortNullDate }`

In my project, this works for both directions, ASC and DESC. Please try it yourself. I hope it helps :smiley:

**Comment by dalbir** [comment]
sortFn: function(a,b){return a > b} (see Sorting and Filtering) It may be renamed to 'sortingAlgorithm'.

**Comment by goleafs** [comment]
this does not seem to work when grouping by that column
the null values still go to the bottom in ASC

**Comment by a3aakash** [comment]
I need the same thing in my project but needs null values on top when the column is in ascending order and null values on the bottom when the column is in descending order. Need functionality similar to Mysql query sorting.

**Comment by cyimking** [comment]
Not sure if this helps out or anything, but this is the custom compare function I used to sort columns with null or emptied values:
`private static compare(v1, v2) {
    if(v1 < v2 || !v1) {
      return -1;
    }

    if(v1 > v2 || !v2) {
      return 1;
    }

    return 0;
  }`

**Comment by AsifSajid** [comment]
@cyimking Thank you, man! I was missing this '||' condition. It did the trick for me.
